File: 104-maximum-depth-of-binary-tree/maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# # //dfs
# class Solution:
#     def maxDepth(self, root: Optional[TreeNode]) -> int:
#         if not root:
#             return 0 
#         return 1 + max(self.maxDepth(root.left), self.maxDepth(root.right))

# //bfs


class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        lvl=0
        q=deque([root])
        while q:
            for i in range(len(q)):
                node=q.popleft()
                if(node.left):
                    q.append(node.left)
                if(node.right):
                    q.append(node.right)
            lvl+=1
        return lvl

File: 104-maximum-depth-of-binary-tree/maximum-depth-of-binary-tree.cpp
// //O(N)
// class Solution {
// public:
//     int maxDepth(TreeNode* root) {
//         if(root==NULL) return 0;
//         int lh=maxDepth(root->left);
//         int rh=maxDepth(root->right);
//         return 1+max(lh,rh);
//     }
// };

// // BFS
class Solution {
public:
int maxDepth(TreeNode *root){
    if(root == NULL)
        return 0;
    int res = 0;
    queue<TreeNode *> q;
    q.push(root);
    while(!q.empty()){
        res++;
        int n = q.size();
        for(int i=0; i < n; i++){
            TreeNode *node = q.front();
            q.pop();
            if(node -> left != NULL)
                q.push(node -> left);
            if(node -> right != NULL)
                q.push(node -> right);
        }
    }
    return res;
}
};